File: examples/AES-CBC-PKCS5Padding/acp.go
package main

import (
    "crypto/aes"
    "crypto/cipher"
    "bytes"
    //"encoding/base64"
    "encoding/hex"
    "fmt"
)

var key = []byte("secretsecretsecr")

var iv = []byte("secretsecretsecr")

type AES_CBC struct {

}

func main(){
    plainText := []byte("Message")
    fmt.Println("plain:", string(plainText))

    //加密串
    cipherText,_ := Encrypt(plainText)
    fmt.Println("cipher:", cipherText)

    //解密串
    newPlain,_ := Decrypt(cipherText)
    fmt.Println("new plain:", newPlain)
}

func Encrypt(origData []byte) (string, error) {
    block, err := aes.NewCipher(key)
    if err != nil {
        return "", err
    }
    blockSize := block.BlockSize()
    origData = PKCS5Padding(origData, blockSize)
    // origData = ZeroPadding(origData, block.BlockSize())
    blockMode := cipher.NewCBCEncrypter(block, iv)
    crypted := make([]byte, len(origData))

    blockMode.CryptBlocks(crypted, origData)
    //输出16进制
    return hex.EncodeToString(crypted),nil
    //输出base64编码
    //return base64.StdEncoding.EncodeToString(crypted), nil
}

func Decrypt(crypted string) (string, error) {
    //16进制时用此方法解码
    decodeData, err := hex.DecodeString(crypted)
    //base64编码时用此方法解码
    //decodeData,err:=base64.StdEncoding.DecodeString(crypted)
    if err != nil {
        return "",err
    }
    block, err := aes.NewCipher(key)
    if err != nil {
        return "", err
    }
    //blockSize := block.BlockSize()
    blockMode := cipher.NewCBCDecrypter(block, iv)
    origData := make([]byte, len(decodeData))
    blockMode.CryptBlocks(origData, decodeData)
    origData = PKCS5UnPadding(origData)
    // origData = ZeroUnPadding(origData)
    return string(origData), nil
}

func ZeroPadding(ciphertext []byte, blockSize int) []byte {
    padding := blockSize - len(ciphertext) % blockSize
    padtext := bytes.Repeat([]byte{0}, padding)
    return append(ciphertext, padtext...)
}

func ZeroUnPadding(origData []byte) []byte {
    length := len(origData)
    unpadding := int(origData[length - 1])
    return origData[:(length - unpadding)]
}

func PKCS5Padding(ciphertext []byte, blockSize int) []byte {
    padding := blockSize - len(ciphertext) % blockSize
    padtext := bytes.Repeat([]byte{byte(padding)}, padding)
    return append(ciphertext, padtext...)
}

func PKCS5UnPadding(origData []byte) []byte {
    length := len(origData)
    // 去掉最后一个字节 unpadding 次
    unpadding := int(origData[length - 1])
    return origData[:(length - unpadding)]
}

File: examples/AES-CBC-PKCS5Padding/acp.py
# coding:utf8
"""
本文采用的AES CBC PKCS5padding方式加解密：

1. 密钥和IV相同，可以自己定义IV的值；

2. 这里加解密接收的数据是字符串 字符串 字符串 ，重要的事情说三遍

3. 当前输出的方式HEX方式，可以使用base64（注掉的那部分）

"""
import re
import base64
import binascii
from Crypto.Cipher import AES


class AESCBC:
    def __init__(self):
        self.key = 'secretsecretsecr'  # 定义key值，16位字符串
        self.iv = self.key[:16]  # iv设置为key的前16位，
        self.mode = AES.MODE_CBC
        self.bs = 16  # block size
        self.PADDING = lambda s: s + (self.bs - len(s) % self.bs) * chr(self.bs - len(s) % self.bs)

    def encrypt(self, text):
        generator = AES.new(self.key, self.mode, self.iv)  # 这里的key 和IV 一样 ，可以按照自己的值定义
        crypt = generator.encrypt(self.PADDING(text))
        # crypted_str = base64.b64encode(crypt)   #输出Base64
        crypted_str = binascii.b2a_hex(crypt)  # 输出Hex
        result = crypted_str.decode()
        return result

    def decrypt(self, text):
        generator = AES.new(self.key, self.mode, self.iv)
        text += (len(text) % 4) * '='
        # decrpyt_bytes = base64.b64decode(text)           #输出Base64
        decrpyt_bytes = binascii.a2b_hex(text)  # 输出Hex
        meg = generator.decrypt(decrpyt_bytes)
        # 去除解码后的非法字符
        try:
            result = re.compile('[\\x00-\\x08\\x0b-\\x0c\\x0e-\\x1f\n\r\t]').sub('', meg.decode())
        except Exception:
            result = '解码失败，请重试!'
        return result


if __name__ == '__main__':
    aes = AESCBC()

    to_encrypt = 'Message'
    to_decrypt = '009d56f792cc5168ee5f31cfda5a8594'
    # base64      'AJ1W95LMUWjuXzHP2lqFlA=='

    print("\n加密前:{0}\n加密后：{1}\n".format(to_encrypt, aes.encrypt(to_encrypt)))
    print("解密前:{0}\n解密后：{1}".format(to_decrypt, aes.decrypt(to_decrypt)))
